rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,770 @@
+only has access to state after the call completes, which will be different if
+the call mutates memory.
+
+Returning to our `pop` function from before we may wish to describe in which
+case the result is `Some`. However that depends on whether `self` is empty
+*before* `pop` is called. To do this Kani provides the `old(EXPR)` pseudo
+function (see [this section](#open-questions) about a discussion on naming),
+which evaluates `EXPR` before the call (e.g. to `pop`) and makes the result
+available to `ensures`. It is used like so:
+
+```rs
+impl<T> Vec<T> {
+  #[kani::ensures(old(self.is_empty()) || result.is_some())]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+`old` allows evaluating any (side-effect free, see [here](#changes-to-other-components)) Rust expression.
+The borrow checker enforces the result of `old` cannot observe the mutations
+from e.g. `pop`, as that would defeat the purpose. If `your` expression in `old`
+returns borrowed content, make a copy instead (using e.g. `clone()`).
+
+Note also that `old` is syntax, not a function and implemented as an extraction
+and lifting during code generation. It can reference e.g. `pop`'s arguments but
+not local variables. Compare the following
+
+**Invalid ❌:** `#[kani::ensures({ let x = self.is_empty(); old(x) } || result.is_some())]`</br>
+**Valid ✅:** `#[kani::ensures(old({ let x = self.is_empty(); x }) || result.is_some())]`
+
+And it will only be recognized as `old(...)`, not as `let old1 = old; old1(...)` etc.
+
+### Workflow and Attribute Contraints Overview
+
+1. By default `kani` or `cargo kani` first verifies all contract harnesses
+   (`proof_for_contract`) reachable from the file or in the local workspace
+   respectively.
+2. Each contract that is used in a `stub_verified` is required to have at least
+   one associated contract harness. Kani reports any missing contract harnesses
+   as errors.
+3. Kani verifies all non-contract harnesses where every `stub_verified` they use
+   had at least one contract harness present and all verifications of contract
+   harnesses succeeded.
+
+When specific harnesses are selected (with `--harness`) contracts are not
+verified.
+
+Kani reports a compile time error if any of the following constraints are violated:
+
+- A function may have any number of `requires`, `ensures`, `modifies` and `frees`
+  attributes. Any function with at least one such annotation is considered as
+  "having a contract".
+
+  Harnesses (general or for contract checking) may not have any such annotation.
+
+- A harness may have up to one `proof_for_contract(TARGET)` annotation where `TARGET` must
+  "have a contract". One or more `proof_for_contract` harnesses may have the
+  same `TARGET`. All such harnesses must pass verification, before `TARGET` may
+  be used as a verified stub.
+
+  A `proof_for_contract` harness may use any harness attributes, including
+  `stub` and `stub_verified`, though the `TARGET` may not appear in either. 
```

Comment:
I think `stub` in general should come with a big red flag, but once you're aware of the dangers I don't see a reason to prevent people from using it. It's not like the unsoundness of using it in conjunction with contracts is somehow higher than using it on its own.